In [2]:
import torch
import gc
torch.cuda.empty_cache()
gc.collect()


0

In [1]:
from setup import setup_src_path
print(setup_src_path())
import data.processed as processed
import config.config as config
import utils.setup as setup
import utils.functions as fn
from importlib import reload



['/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/mlm', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '', '/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages', '/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/modules']


2024-06-04 03:27:23.150332: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-06-04 03:27:23.183065: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-06-04 03:27:23.700313: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


In [2]:
from datasets import load_from_disk,concatenate_datasets


adapter_name="mlm_union_TE"
dataset = load_from_disk(f"{config.Config.DATASETS_SAVE_PATH}/datasets")



In [3]:
reload(config)

    
filtered_target = dataset['train'].filter(lambda example: example['genre'] == "telephone").shuffle(seed=42)
#filtered_target = shuffled_filtered_target.train_test_split(test_size=0.1)


filtered_test_target = dataset['validation_matched'].filter(lambda example: example['genre'] == "telephone")
train_target = filtered_target
test_target = filtered_test_target

train_target[2]

{'promptID': 8103,
 'pairID': '8103e',
 'premise': "but i like dogs and my husband like cats so we haven't reached a real agreement on that yet  if we get a place where we can have both it'll be great but until then",
 'premise_binary_parse': "( but ( ( i ( like ( ( dogs and ) ( ( my husband ) ( like cats ) ) ) ) ) ( so ( we ( ( have n't ) ( reached ( ( a ( real agreement ) ) ( ( on that ) ( ( yet ( if ( we ( get ( ( a place ) ( where ( we ( can ( have both ) ) ) ) ) ) ) ) ) ( it ( 'll ( be ( great ( but ( until then ) ) ) ) ) ) ) ) ) ) ) ) ) ) )",
 'premise_parse': "(ROOT (FRAG (CC but) (ADVP (RB i) (PP (IN like) (NP (NP (NNS dogs)) (CC and) (NP (NP (PRP$ my) (NN husband)) (PP (IN like) (NP (NNS cats))))))) (SBAR (IN so) (S (NP (PRP we)) (VP (VBP have) (RB n't) (VP (VBN reached) (NP (NP (DT a) (JJ real) (NN agreement)) (SBAR (WHPP (IN on) (WHNP (WDT that))) (S (SBAR (RB yet) (IN if) (S (NP (PRP we)) (VP (VBP get) (NP (NP (DT a) (NN place)) (SBAR (WHADVP (WRB where)) (S (NP (PRP we)) (

In [4]:
from adapters import AutoAdapterModel,init
from tqdm import tqdm

from transformers import AutoConfig, AutoModelForMaskedLM

mdlcfg = AutoConfig.from_pretrained(
    config.Config.MODEL_NAME,
 
)
model = AutoModelForMaskedLM.from_pretrained(
    config.Config.MODEL_NAME,
)
init(model)
reload(fn)
fn.print_trainable_parameters(model)

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertFor

trainable params: 109514298 || all params: 109514298 || trainable%: 100.0


In [5]:
from adapters import ConfigUnion, LoRAConfig, PrefixTuningConfig, SeqBnConfig,SeqBnInvConfig,AdapterConfig,LoRAConfig

union_config = ConfigUnion(
    SeqBnInvConfig(reduction_factor=2),
    LoRAConfig(r=8, alpha=16),
)
model.add_adapter(adapter_name, config=union_config)

model.train_adapter([adapter_name])
model.active_adapters = adapter_name
print(model.adapter_summary())
fn.print_trainable_parameters(model)

Name                     Architecture         #Param      %Param  Active   Train
--------------------------------------------------------------------------------
mlm_union_TE             union             7,682,688       7.055       1       1
--------------------------------------------------------------------------------
Full model                               108,891,648     100.000               0
trainable params: 8274816 || all params: 117196986 || trainable%: 7.060604783812444


In [6]:
from transformers import AutoTokenizer

reload(processed)
tokenizer = AutoTokenizer.from_pretrained(config.Config.TOKENIZER_NAME)

tokenized_dataset= processed.tokenize_dataset(train_target,tokenizer)
tokenized_dataset

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids'],
    num_rows: 83348
})

In [7]:
chunk_size = 128
# Slicing produces a list of lists for each feature
tokenized_samples = tokenized_dataset[444:470]

results = fn.group_texts(tokenized_samples, chunk_size)
for chunk in results["labels"]:
    print(f"'>>> Chunk length: {len(chunk)}'")

'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'


In [8]:
# we group texts and chunk them
lm_datasets = tokenized_dataset.map(fn.group_texts, batched=True,fn_kwargs={'chunk_size': chunk_size})
lm_datasets

Map:   0%|          | 0/83348 [00:00<?, ? examples/s]

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
    num_rows: 27551
})

In [9]:
from transformers import DataCollatorForLanguageModeling

data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm_probability=0.15)
samples = [lm_datasets[i] for i in range(2)]
for sample in samples:
    _ = sample.pop("word_ids")
dd=data_collator(samples)
for chunk in dd["input_ids"]:
    print(f"\n'>>> {tokenizer.decode(chunk)}'")


'>>> [CLS] i [MASK] i was absolutely flabbergast [MASK] at what she they would that she was paid true she was [MASK] you [MASK] but still it's the principle of course if she got her [MASK] [MASK] s degree it'[MASK] all [MASK] different but uh [SEP] i think she deserves much more money. [SEP] [CLS] yeah and you knowlion it there are many of themfahan [MASK] s no doubt about it [SEP] there are countless [MASK] of them, [MASK] is for sure. [SEP] [CLS] but i like dogs and my husband like cats so we haven't reached a real [MASK] on that yet if we get a place where we can have both it ''

'>>> ll be great but until then [SEP] my [MASK] and i can't agree on a pet [MASK] [SEP] [CLS] yeah tell me about it tell me [MASK] lucky husband's putting two of [MASK] [MASK] isn't he lucky [SEP] my husband is lucky because [MASK] is putting two of [MASK] through [MASK] [SEP] [CLS] never breaks down um the styrofoam [SEP] never destroys the [MASK]yrof [MASK] [MASK]. [SEP] [CLS] [MASK] haven't seen that an

In [10]:

downsampled_dataset = lm_datasets.train_test_split(
    test_size=0.1, seed=42
)
downsampled_dataset

DatasetDict({
    train: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 24795
    })
    test: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 2756
    })
})

In [11]:

reload(fn)
trainer = fn.train_mlm_model(model,adapter_name,data_collator,tokenizer, downsampled_dataset['train'],downsampled_dataset['test'])

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/accelerate/accelerator.py:446: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(


In [12]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/87 [00:00<?, ?it/s]

>>> Perplexity: 23.69


In [13]:
trainer.train()

  0%|          | 0/15500 [00:00<?, ?it/s]

{'loss': 2.3552, 'grad_norm': 1.8179911375045776, 'learning_rate': 9.817333333333334e-05, 'epoch': 1.0}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.9778926372528076, 'eval_runtime': 12.0154, 'eval_samples_per_second': 229.373, 'eval_steps_per_second': 7.241, 'epoch': 1.0}
{'loss': 2.0178, 'grad_norm': 1.607722282409668, 'learning_rate': 9.301333333333335e-05, 'epoch': 2.0}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.8736953735351562, 'eval_runtime': 12.0394, 'eval_samples_per_second': 228.916, 'eval_steps_per_second': 7.226, 'epoch': 2.0}
{'loss': 1.9372, 'grad_norm': 1.6166468858718872, 'learning_rate': 8.785333333333335e-05, 'epoch': 3.0}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.829020380973816, 'eval_runtime': 12.1847, 'eval_samples_per_second': 226.186, 'eval_steps_per_second': 7.14, 'epoch': 3.0}
{'loss': 1.9024, 'grad_norm': 1.7206645011901855, 'learning_rate': 8.269333333333334e-05, 'epoch': 3.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.8137454986572266, 'eval_runtime': 12.0156, 'eval_samples_per_second': 229.368, 'eval_steps_per_second': 7.241, 'epoch': 4.0}
{'loss': 1.8676, 'grad_norm': 1.744458794593811, 'learning_rate': 7.753333333333334e-05, 'epoch': 4.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7581197023391724, 'eval_runtime': 8.4389, 'eval_samples_per_second': 326.584, 'eval_steps_per_second': 10.309, 'epoch': 5.0}
{'loss': 1.8374, 'grad_norm': 1.6628146171569824, 'learning_rate': 7.237333333333334e-05, 'epoch': 5.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7917754650115967, 'eval_runtime': 8.4518, 'eval_samples_per_second': 326.086, 'eval_steps_per_second': 10.294, 'epoch': 6.0}
{'loss': 1.8195, 'grad_norm': 1.7069979906082153, 'learning_rate': 6.721333333333334e-05, 'epoch': 6.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7675310373306274, 'eval_runtime': 8.4738, 'eval_samples_per_second': 325.237, 'eval_steps_per_second': 10.267, 'epoch': 7.0}
{'loss': 1.8, 'grad_norm': 1.7157156467437744, 'learning_rate': 6.206000000000001e-05, 'epoch': 7.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7280482053756714, 'eval_runtime': 8.4734, 'eval_samples_per_second': 325.254, 'eval_steps_per_second': 10.267, 'epoch': 8.0}
{'loss': 1.7799, 'grad_norm': 1.796109676361084, 'learning_rate': 5.69e-05, 'epoch': 8.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.738039255142212, 'eval_runtime': 8.479, 'eval_samples_per_second': 325.039, 'eval_steps_per_second': 10.261, 'epoch': 9.0}
{'loss': 1.7685, 'grad_norm': 1.7790781259536743, 'learning_rate': 5.174e-05, 'epoch': 9.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6871061325073242, 'eval_runtime': 8.4332, 'eval_samples_per_second': 326.804, 'eval_steps_per_second': 10.316, 'epoch': 10.0}
{'loss': 1.7601, 'grad_norm': 1.734527587890625, 'learning_rate': 4.6586666666666666e-05, 'epoch': 10.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6967809200286865, 'eval_runtime': 4.7476, 'eval_samples_per_second': 580.499, 'eval_steps_per_second': 18.325, 'epoch': 11.0}
{'loss': 1.7462, 'grad_norm': 1.7619233131408691, 'learning_rate': 4.142666666666667e-05, 'epoch': 11.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.711893916130066, 'eval_runtime': 4.7262, 'eval_samples_per_second': 583.131, 'eval_steps_per_second': 18.408, 'epoch': 12.0}
{'loss': 1.7264, 'grad_norm': 1.7179824113845825, 'learning_rate': 3.626666666666667e-05, 'epoch': 12.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6649487018585205, 'eval_runtime': 4.7283, 'eval_samples_per_second': 582.872, 'eval_steps_per_second': 18.4, 'epoch': 13.0}
{'loss': 1.7287, 'grad_norm': 1.6768529415130615, 'learning_rate': 3.111333333333333e-05, 'epoch': 13.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6588877439498901, 'eval_runtime': 4.7348, 'eval_samples_per_second': 582.069, 'eval_steps_per_second': 18.374, 'epoch': 14.0}
{'loss': 1.7122, 'grad_norm': 1.7260767221450806, 'learning_rate': 2.5960000000000002e-05, 'epoch': 14.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6429065465927124, 'eval_runtime': 4.7476, 'eval_samples_per_second': 580.509, 'eval_steps_per_second': 18.325, 'epoch': 15.0}
{'loss': 1.7059, 'grad_norm': 1.7947477102279663, 'learning_rate': 2.08e-05, 'epoch': 15.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6447105407714844, 'eval_runtime': 4.7239, 'eval_samples_per_second': 583.419, 'eval_steps_per_second': 18.417, 'epoch': 16.0}
{'loss': 1.6958, 'grad_norm': 1.8761253356933594, 'learning_rate': 1.5640000000000003e-05, 'epoch': 16.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6680830717086792, 'eval_runtime': 4.7281, 'eval_samples_per_second': 582.895, 'eval_steps_per_second': 18.401, 'epoch': 17.0}
{'loss': 1.6913, 'grad_norm': 1.9000506401062012, 'learning_rate': 1.0480000000000001e-05, 'epoch': 17.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.648944616317749, 'eval_runtime': 4.7416, 'eval_samples_per_second': 581.238, 'eval_steps_per_second': 18.348, 'epoch': 18.0}


Overwriting existing adapter 'mlm_union_TE'.


{'train_runtime': 3002.0963, 'train_samples_per_second': 165.185, 'train_steps_per_second': 5.163, 'train_loss': 1.8249022794880747, 'epoch': 18.0}


TrainOutput(global_step=13950, training_loss=1.8249022794880747, metrics={'train_runtime': 3002.0963, 'train_samples_per_second': 165.185, 'train_steps_per_second': 5.163, 'train_loss': 1.8249022794880747, 'epoch': 18.0})

In [14]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/87 [00:00<?, ?it/s]

>>> Perplexity: 5.11


In [15]:
import config.config as config

trainer.model.save_adapter(f"{config.Config.ADAPTER_SAVE_PATH}/{adapter_name}", adapter_name)


: 

#### This is a report of mlm for goverment domain using whole genre as target and then splitt after tokenization with seed 42. I used computer metrix for preplexity and early stopping 
The results before:
>>> Perplexity: 23.69

The results after:
>>> Perplexity: 5.11